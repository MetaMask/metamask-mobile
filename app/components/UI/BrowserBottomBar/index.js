import React, { PureComponent } from 'react';
import { Platform, TouchableOpacity, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';
import ElevatedView from 'react-native-elevated-view';
import TabCountIcon from '../Tabs/TabCountIcon';
import Icon from 'react-native-vector-icons/FontAwesome';
import MaterialIcon from 'react-native-vector-icons/MaterialIcons';
import SimpleLineIcons from 'react-native-vector-icons/SimpleLineIcons';
import FeatherIcons from 'react-native-vector-icons/Feather';
import { MetaMetricsEvents } from '../../../core/Analytics';
import AnalyticsV2 from '../../../util/analyticsV2';

import Device from '../../../util/device';
import { ThemeContext, mockTheme } from '../../../util/theme';

// NOTE: not needed anymore. The use of BottomTabBar already accomodates the home indicator height
// TODO: test on an android device
// const HOME_INDICATOR_HEIGHT = 0;
// const defaultBottomBarPadding = 0;

const createStyles = (colors) =>
  StyleSheet.create({
    bottomBar: {
      backgroundColor: colors.background.default,
      flexDirection: 'row',
      flex: 0,
      borderTopWidth: Device.isAndroid() ? 0 : StyleSheet.hairlineWidth,
      borderColor: colors.border.muted,
      justifyContent: 'space-between',
    },
    iconButton: {
      height: 24,
      width: 24,
      justifyContent: 'space-around',
      alignItems: 'center',
      textAlign: 'center',
      flex: 1,
      paddingTop: 30,
      paddingBottom: 30,
    },
    tabIcon: {
      marginTop: 0,
      width: 24,
      height: 24,
    },
    disabledIcon: {
      color: colors.icon.muted,
    },
    icon: {
      width: 24,
      height: 24,
      color: colors.icon.default,
      textAlign: 'center',
    },
  });

/**
 * Browser bottom bar that contains icons for navigation
 * tab management, url change and other options
 */
export default class BrowserBottomBar extends PureComponent {
  static propTypes = {
    /**
     * Boolean that determines if you can navigate back
     */
    canGoBack: PropTypes.bool,
    /**
     * Boolean that determines if you can navigate forward
     */
    canGoForward: PropTypes.bool,
    /**
     * Function that allows you to navigate back
     */
    goBack: PropTypes.func,
    /**
     * Function that allows you to navigate forward
     */
    goForward: PropTypes.func,
    /**
     * Function that triggers the tabs view
     */
    showTabs: PropTypes.func,
    /**
     * Function that triggers the change url modal view
     */
    showUrlModal: PropTypes.func,
    /**
     * Function that redirects to the home screen
     */
    goHome: PropTypes.func,
    /**
     * Function that toggles the options menu
     */
    toggleOptions: PropTypes.func,
  };

  trackSearchEvent = () => {
    AnalyticsV2.trackEvent(MetaMetricsEvents.BROWSER_SEARCH_USED, {
      option_chosen: 'Browser Bottom Bar Menu',
      number_of_tabs: undefined,
    });
  };

  trackNavigationEvent = (navigationOption) => {
    AnalyticsV2.trackEvent(MetaMetricsEvents.BROWSER_NAVIGATION, {
      option_chosen: navigationOption,
      os: Platform.OS,
    });
  };

  render() {
    const {
      canGoBack,
      goBack,
      canGoForward,
      goForward,
      showTabs,
      goHome,
      showUrlModal,
      toggleOptions,
    } = this.props;
    const colors = this.context.colors || mockTheme.colors;
    const styles = createStyles(colors);

    const onSearchPress = () => {
      showUrlModal();
      this.trackSearchEvent();
    };

    const onBackPress = () => {
      goBack();
      this.trackNavigationEvent('Go Back');
    };

    const onForwardPress = () => {
      goForward();
      this.trackNavigationEvent('Go Forward');
    };

    const onHomePress = () => {
      goHome();
      this.trackNavigationEvent('Go Home');
    };

    return (
      <ElevatedView elevation={11} style={styles.bottomBar}>
        <TouchableOpacity
          onPress={onBackPress}
          style={styles.iconButton}
          testID={'go-back-button'}
          disabled={!canGoBack}
        >
          <Icon
            name="angle-left"
            size={24}
            style={[styles.icon, !canGoBack ? styles.disabledIcon : {}]}
          />
        </TouchableOpacity>
        <TouchableOpacity
          onPress={onForwardPress}
          style={styles.iconButton}
          testID={'go-forward-button'}
          disabled={!canGoForward}
        >
          <Icon
            name="angle-right"
            size={24}
            style={[styles.icon, !canGoForward ? styles.disabledIcon : {}]}
          />
        </TouchableOpacity>
        <TouchableOpacity
          onPress={onSearchPress}
          style={styles.iconButton}
          testID={'search-button'}
        >
          <FeatherIcons name="search" size={24} style={styles.icon} />
        </TouchableOpacity>

        <TouchableOpacity
          onPress={showTabs}
          style={styles.iconButton}
          testID={'show-tabs-button'}
        >
          <TabCountIcon style={styles.tabIcon} />
        </TouchableOpacity>
        <TouchableOpacity
          onPress={onHomePress}
          style={styles.iconButton}
          testID={'home-button'}
        >
          <SimpleLineIcons name="home" size={22} style={styles.icon} />
        </TouchableOpacity>

        <TouchableOpacity
          onPress={toggleOptions}
          style={styles.iconButton}
          testID={'options-button'}
        >
          <MaterialIcon name="more-horiz" size={22} style={styles.icon} />
        </TouchableOpacity>
      </ElevatedView>
    );
  }
}

BrowserBottomBar.contextType = ThemeContext;
