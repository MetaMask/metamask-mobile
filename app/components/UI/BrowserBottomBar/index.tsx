import React, { useContext } from 'react';
import { Platform, TouchableOpacity, StyleSheet } from 'react-native';
import ElevatedView from 'react-native-elevated-view';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import TabCountIcon from '../Tabs/TabCountIcon';
import Icon from 'react-native-vector-icons/FontAwesome';
import MaterialIcon from 'react-native-vector-icons/MaterialIcons';
import SimpleLineIcons from 'react-native-vector-icons/SimpleLineIcons';
import FeatherIcons from 'react-native-vector-icons/Feather';
import { MetaMetricsEvents } from '../../../core/Analytics';

import Device from '../../../util/device';
import { ThemeContext, mockTheme } from '../../../util/theme';
import { BrowserViewSelectorsIDs } from '../../../../e2e/selectors/Browser/BrowserView.selectors';
import { useMetrics } from '../../../components/hooks/useMetrics';
import { ThemeColors } from '@metamask/design-tokens';

// NOTE: not needed anymore. The use of BottomTabBar already accomodates the home indicator height
// TODO: test on an android device
// const HOME_INDICATOR_HEIGHT = 0;
// const defaultBottomBarPadding = 0;

const createStyles = (colors: ThemeColors, tabBarBottomInset = 0) =>
  StyleSheet.create({
    bottomBar: {
      backgroundColor: colors.background.default,
      flexDirection: 'row',
      flex: 0,
      borderTopWidth: Device.isAndroid() ? 0 : StyleSheet.hairlineWidth,
      borderColor: colors.border.muted,
      justifyContent: 'space-between',
      paddingBottom: tabBarBottomInset,
    },
    iconButton: {
      height: 60,
      width: 24,
      justifyContent: 'space-around',
      alignItems: 'center',
      textAlign: 'center',
      flex: 1,
    },
    tabIcon: {
      marginTop: 0,
      width: 24,
      height: 24,
    },
    disabledIcon: {
      color: colors.icon.muted,
    },
    icon: {
      width: 24,
      height: 24,
      color: colors.icon.default,
      textAlign: 'center',
    },
  });

interface BrowserBottomBarProps {
  /**
   * Boolean that determines if you can navigate back
   */
  canGoBack?: boolean;
  /**
   * Boolean that determines if you can navigate forward
   */
  canGoForward?: boolean;
  /**
   * Function that allows you to navigate back
   */
  goBack?: () => void;
  /**
   * Function that allows you to navigate forward
   */
  goForward?: () => void;
  /**
   * Function that triggers the tabs view
   */
  showTabs?: () => void;
  /**
   * Function that triggers the change url modal view
   */
  showUrlModal?: () => void;
  /**
   * Function that redirects to the home screen
   */
  goHome?: () => void;
  /**
   * Function that toggles the options menu
   */
  toggleOptions?: () => void;
  /**
   * Function that toggles fullscreen mode
   */
  toggleFullscreen?: () => void;
  /**
   * Boolean that determines if currently in fullscreen mode
   */
  isFullscreen?: boolean;
}

/**
 * Browser bottom bar that contains icons for navigation
 * tab management, url change and other options
 */
const BrowserBottomBar: React.FC<BrowserBottomBarProps> = ({
  canGoBack,
  canGoForward,
  goBack,
  goForward,
  showTabs,
  showUrlModal,
  goHome,
  toggleOptions,
  toggleFullscreen,
  isFullscreen,
}) => {
  const { bottom: bottomInset } = useSafeAreaInsets();
  const { colors } = useContext(ThemeContext) || { colors: mockTheme.colors };
  const { trackEvent, createEventBuilder } = useMetrics();

  const styles = createStyles(colors, isFullscreen ? bottomInset : 0);

  const trackSearchEvent = (): void => {
    trackEvent(
      createEventBuilder(MetaMetricsEvents.BROWSER_SEARCH_USED)
        .addProperties({
          option_chosen: 'Browser Bottom Bar Menu',
          number_of_tabs: undefined,
        })
        .build(),
    );
  };

  const trackNavigationEvent = (navigationOption: string): void => {
    trackEvent(
      createEventBuilder(MetaMetricsEvents.BROWSER_NAVIGATION)
        .addProperties({
          option_chosen: navigationOption,
          os: Platform.OS,
        })
        .build(),
    );
  };

  const onSearchPress = (): void => {
    showUrlModal?.();
    trackSearchEvent();
  };

  const onBackPress = (): void => {
    goBack?.();
    trackNavigationEvent('Go Back');
  };

  const onForwardPress = (): void => {
    goForward?.();
    trackNavigationEvent('Go Forward');
  };

  const onHomePress = (): void => {
    goHome?.();
    trackNavigationEvent('Go Home');
  };

  const onToggleFullscreenPress = (): void => {
    if (isFullscreen) {
      trackEvent(
        createEventBuilder(MetaMetricsEvents.BROWSER_CLOSED_FULLSCREEN).build(),
      );
    } else {
      trackEvent(
        createEventBuilder(MetaMetricsEvents.BROWSER_OPENED_FULLSCREEN).build(),
      );
    }
    toggleFullscreen?.();
  };

  const homeDisabled = !goHome;
  const optionsDisabled = !toggleOptions;

  return (
    <ElevatedView elevation={11} style={styles.bottomBar}>
      <TouchableOpacity
        onPress={onBackPress}
        style={styles.iconButton}
        testID={BrowserViewSelectorsIDs.BACK_BUTTON}
        disabled={!canGoBack}
      >
        <Icon
          name="angle-left"
          size={24}
          style={[styles.icon, !canGoBack ? styles.disabledIcon : {}]}
        />
      </TouchableOpacity>
      <TouchableOpacity
        onPress={onForwardPress}
        style={styles.iconButton}
        testID={BrowserViewSelectorsIDs.FORWARD_BUTTON}
        disabled={!canGoForward}
      >
        <Icon
          name="angle-right"
          size={24}
          style={[styles.icon, !canGoForward ? styles.disabledIcon : {}]}
        />
      </TouchableOpacity>
      <TouchableOpacity
        onPress={onSearchPress}
        style={styles.iconButton}
        testID={BrowserViewSelectorsIDs.SEARCH_BUTTON}
      >
        <FeatherIcons name="search" size={24} style={styles.icon} />
      </TouchableOpacity>
      <TouchableOpacity
        onPress={showTabs}
        style={styles.iconButton}
        testID={BrowserViewSelectorsIDs.TABS_BUTTON}
      >
        <TabCountIcon style={styles.tabIcon} />
      </TouchableOpacity>
      <TouchableOpacity
        onPress={onHomePress}
        style={styles.iconButton}
        testID={BrowserViewSelectorsIDs.HOME_BUTTON}
        disabled={homeDisabled}
      >
        <SimpleLineIcons
          name="home"
          size={22}
          style={[styles.icon, homeDisabled && styles.disabledIcon]}
        />
      </TouchableOpacity>
      <TouchableOpacity
        onPress={onToggleFullscreenPress}
        style={styles.iconButton}
        testID={BrowserViewSelectorsIDs.TOGGLE_FULLSCREEN_BUTTON}
        disabled={!toggleFullscreen}
      >
        <MaterialIcon
          name={isFullscreen ? 'fullscreen-exit' : 'fullscreen'}
          size={22}
          style={[styles.icon, !toggleFullscreen ? styles.disabledIcon : {}]}
        />
      </TouchableOpacity>
      <TouchableOpacity
        onPress={toggleOptions}
        style={styles.iconButton}
        testID={BrowserViewSelectorsIDs.OPTIONS_BUTTON}
        disabled={optionsDisabled}
      >
        <MaterialIcon
          name="more-horiz"
          size={22}
          style={[styles.icon, optionsDisabled && styles.disabledIcon]}
        />
      </TouchableOpacity>
    </ElevatedView>
  );
};

export default React.memo(BrowserBottomBar);
