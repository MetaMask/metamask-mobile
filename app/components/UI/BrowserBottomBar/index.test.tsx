/* eslint-disable react/jsx-no-bind */
import React from 'react';
import { shallow } from 'enzyme';
import BrowserBottomBar from './';

describe('BrowserBottomBar', () => {
  it('should render correctly', () => {
    const fn = () => null;

    const wrapper = shallow(
      <BrowserBottomBar
        canGoBack
        canGoForward={false}
        showTabs={fn}
        toggleOptions={fn}
        showUrlModal={fn}
        goBack={fn}
        goForward={fn}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
