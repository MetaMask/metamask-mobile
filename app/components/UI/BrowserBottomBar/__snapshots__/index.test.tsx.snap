// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowserBottomBar should render correctly 1`] = `
<ElevatedView
  elevation={11}
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "borderColor": "#D6D9DC",
      "borderTopWidth": 0.5,
      "flex": 0,
      "flexDirection": "row",
      "justifyContent": "space-between",
    }
  }
>
  <TouchableOpacity
    disabled={false}
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "height": 24,
        "justifyContent": "space-around",
        "paddingBottom": 30,
        "paddingTop": 30,
        "textAlign": "center",
        "width": 24,
      }
    }
    testID="go-back-button"
  >
    <Icon
      allowFontScaling={false}
      name="angle-left"
      size={24}
      style={
        Array [
          Object {
            "color": "#24272A",
            "height": 24,
            "textAlign": "center",
            "width": 24,
          },
          Object {},
        ]
      }
    />
  </TouchableOpacity>
  <TouchableOpacity
    disabled={true}
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "height": 24,
        "justifyContent": "space-around",
        "paddingBottom": 30,
        "paddingTop": 30,
        "textAlign": "center",
        "width": 24,
      }
    }
    testID="go-forward-button"
  >
    <Icon
      allowFontScaling={false}
      name="angle-right"
      size={24}
      style={
        Array [
          Object {
            "color": "#24272A",
            "height": 24,
            "textAlign": "center",
            "width": 24,
          },
          Object {
            "color": "#BBC0C5",
          },
        ]
      }
    />
  </TouchableOpacity>
  <TouchableOpacity
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "height": 24,
        "justifyContent": "space-around",
        "paddingBottom": 30,
        "paddingTop": 30,
        "textAlign": "center",
        "width": 24,
      }
    }
    testID="search-button"
  >
    <Icon
      allowFontScaling={false}
      name="search"
      size={24}
      style={
        Object {
          "color": "#24272A",
          "height": 24,
          "textAlign": "center",
          "width": 24,
        }
      }
    />
  </TouchableOpacity>
  <TouchableOpacity
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "height": 24,
        "justifyContent": "space-around",
        "paddingBottom": 30,
        "paddingTop": 30,
        "textAlign": "center",
        "width": 24,
      }
    }
    testID="show-tabs-button"
  >
    <Connect(TabCountIcon)
      style={
        Object {
          "height": 24,
          "marginTop": 0,
          "width": 24,
        }
      }
    />
  </TouchableOpacity>
  <TouchableOpacity
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "height": 24,
        "justifyContent": "space-around",
        "paddingBottom": 30,
        "paddingTop": 30,
        "textAlign": "center",
        "width": 24,
      }
    }
    testID="home-button"
  >
    <Icon
      allowFontScaling={false}
      name="home"
      size={22}
      style={
        Object {
          "color": "#24272A",
          "height": 24,
          "textAlign": "center",
          "width": 24,
        }
      }
    />
  </TouchableOpacity>
  <TouchableOpacity
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "height": 24,
        "justifyContent": "space-around",
        "paddingBottom": 30,
        "paddingTop": 30,
        "textAlign": "center",
        "width": 24,
      }
    }
    testID="options-button"
  >
    <Icon
      allowFontScaling={false}
      name="more-horiz"
      size={22}
      style={
        Object {
          "color": "#24272A",
          "height": 24,
          "textAlign": "center",
          "width": 24,
        }
      }
    />
  </TouchableOpacity>
</ElevatedView>
`;
